manubot/cite/types.py

```diff
@@ -0,0 +1,158 @@
+from dataclasses import dataclass
+
+
+class CiteKeyParsingError(ValueError):
+    pass
+
+
+def split_prefixed_identifier(string: str):
+    """
+    Split strings like 'doi:blah' into source prefix and identifier 
+    ('doi', 'blah')
+    String must contain a colon (:). Accepts strings starting with '@'.
+    Raises error if source prefix not defined.
+    """
+    string = string.strip()
+    trimmed = string[1:] if string.startswith('@') else string
+    try:
+        source, identifier = trimmed.split(':', 1)
+    except ValueError:
+        raise CiteKeyParsingError(f"Could not process '{string}' as citekey.")
+    source = source.strip().lower()
+    identifier = identifier.strip()
+    Handle.validate(source)  # raises error on wrong source prefix
+    return source, identifier
+
+
+class CiteKey(object):
+    def __init__(self, citekey):
+        self.source, self.identifier = split_prefixed_identifier(citekey)
+
+    def handle(self):
+        return Handle.create_with(self.source, self.identifier)
+
+    def str(self):
+        """Use in chained transformation of citekey as citekey.str()
+           Same result as str(citekey)"""
+        return str(self)
+
+    def __str__(self):
+        return f'{self.source}:{self.identifier}'
+
+    def __repr__(self):
+        return "CiteKey(citekey='{}')".format(self)
+
+
+class HandleBase:
+    """
+    Parent (grandfather) class for handles like DOI, ISBN and other.
+
+    Public classmethods:
+      .source_prefixes() -> [str]
+      .validate(source: str)   
+      .create_with(source: str, identifier: str)
+
+    """
+   
+    @classmethod
+    def _class_dict(cls):
+        return {c.__name__.lower(): c for c in cls.__subclasses__()}
+
+    @classmethod
+    def source_prefixes(cls)  -> [str]:
+        return list(cls._class_dict().keys())
+
+    @classmethod
+    def _get_constructor(cls, key):
+        try:
+            return cls._class_dict()[key.lower()]
+        except KeyError:
+            raise CiteKeyParsingError(
+                    f'Prefix {key} not supported.\n'
+                    f'Must be one of {cls.source_prefixes()}')
+
+    @classmethod
+    def validate(cls, source):
+        _ = cls._get_constructor(source)
+            
+    @classmethod
+    def create_with(cls, source: str, identifer: str):
+        """
+        Create DOI, ISBN and other handles from source name (eg 'doi')
+        and identifier (eg '10.1038/d41586-019-02978-7')
+        """
+        constr = cls._get_constructor(source)
+        return constr(identifer)
+
+
+@dataclass
+class Handle(HandleBase):
+    identifier: str
+
+    # Child classes must redefine canonic() and csl_item()
```

Comment:
Is using `@abc.abstractmethod` a way to enforce this?

Reply:
For `csl_item()` - probably yes, for `canonic()` probably not. `csl_item()` has to change in every child class, while `canonic()` stays the same for all classes except `DOI` and `ISBN`. 
I did not use ABCs before, so unfamiliar with caveats.